File: Vista/CajaPanel.py
from PyQt6.QtWidgets import QWidget, QVBoxLayout, QHBoxLayout, QLabel, QStackedWidget, QPushButton, QLineEdit, QTableWidget, QTableWidgetItem, QHeaderView, QFrame
from PyQt6.QtCore import Qt
from PyQt6.QtGui import QFont, QColor
from Vista.Estilos import Estilos
from Modelo.Movimiento import Movimiento, Tipo as TipoMovimiento
from datetime import datetime

class CajaPanel(QWidget):
    TAB_CERRADA = 0
    TAB_ABIERTA = 1

    def __init__(self, ctrl, usuario, ventana):
        super().__init__()
        self.ctrl = ctrl
        self.usuario = usuario
        self.ventana = ventana
        self.construir()

    def construir(self):
        self.setStyleSheet(f"background-color: {Estilos.BG_CLARO.name()};")
        layout = QVBoxLayout(self)
        layout.setContentsMargins(24, 24, 24, 24)

        self.cardLayout = QStackedWidget()
        self.cardLayout.setStyleSheet(f"background-color: {Estilos.BG_CLARO.name()};")

        self.cardLayout.addWidget(self.construirPanelCerrada())
        self.cardLayout.addWidget(self.construirPanelAbierta())

        layout.addWidget(self.cardLayout)
        self.refrescar()

    def construirPanelCerrada(self):
        p = QWidget()
        p_layout = QVBoxLayout(p)
        p_layout.setAlignment(Qt.AlignmentFlag.AlignCenter)

        tarjeta = QWidget()
        tarjeta.setFixedSize(420, 440)
        tarjeta.setStyleSheet(f"background-color: {Estilos.BG_BLANCO.name()}; border: 1px solid {Estilos.BORDE.name()}; border-radius: 8px;")
        t_layout = QVBoxLayout(tarjeta)
        t_layout.setContentsMargins(44, 40, 44, 40)
        t_layout.setAlignment(Qt.AlignmentFlag.AlignTop)

        icono = QLabel("💵")
        icono.setFont(QFont("SansSerif", 64))
        icono.setAlignment(Qt.AlignmentFlag.AlignCenter)
        icono.setStyleSheet("border: none;")
        t_layout.addWidget(icono)

        t_layout.addSpacing(16)

        tit = QLabel("Terminal Cerrada")
        tit.setFont(Estilos.FUENTE_TITULO)
        tit.setAlignment(Qt.AlignmentFlag.AlignCenter)
        tit.setStyleSheet("border: none;")
        t_layout.addWidget(tit)

        resp = QLabel(f"Responsable: {self.usuario.getNombre()}")
        resp.setFont(Estilos.FUENTE_SMALL)
        resp.setStyleSheet(f"color: {Estilos.TEXTO_SECUNDARIO.name()}; border: none;")
        resp.setAlignment(Qt.AlignmentFlag.AlignCenter)
        t_layout.addWidget(resp)

        t_layout.addSpacing(28)

        lblFondoTexto = QLabel("FONDO INICIAL ($)")
        lblFondoTexto.setFont(Estilos.FUENTE_XS)
        lblFondoTexto.setStyleSheet(f"color: {Estilos.TEXTO_TENUE.name()}; border: none;")
        t_layout.addWidget(lblFondoTexto)

        self.txtFondo = QLineEdit("0.00")
        self.txtFondo.setFont(QFont("SansSerif", 26, QFont.Weight.Bold))
        self.txtFondo.setStyleSheet(f"""
            QLineEdit {{
                background-color: {Estilos.BG_ZINC_100.name()};
                border: 2px solid {Estilos.BORDE.name()};
                padding: 12px 16px;
                border-radius: 4px;
            }}
        """)
        t_layout.addWidget(self.txtFondo)

        t_layout.addSpacing(24)

        btnAbrir = Estilos.botonPrimario("ABRIR TERMINAL")
        btnAbrir.setFont(QFont("SansSerif", 16, QFont.Weight.Bold))
        btnAbrir.setFixedHeight(56)
        btnAbrir.clicked.connect(self.abrirCaja)
        t_layout.addWidget(btnAbrir)

        p_layout.addWidget(tarjeta)
        return p

    def abrirCaja(self):
        try:
            fondo = float(self.txtFondo.text().strip())
            self.ctrl.abrirCaja(fondo)
        except ValueError:
            self.ctrl.abrirCaja(0)

    def construirPanelAbierta(self):
        p = QWidget()
        layout = QVBoxLayout(p)
        layout.setSpacing(16)
        layout.setContentsMargins(0, 0, 0, 0)

        header = QWidget()
        header.setStyleSheet(f"background-color: {Estilos.BG_BLANCO.name()}; border: 1px solid {Estilos.BORDE.name()};")
        h_layout = QHBoxLayout(header)
        h_layout.setContentsMargins(24, 20, 24, 20)

        tit = QLabel("GESTIÓN DE CAJA  —  Operativa")
        tit.setFont(Estilos.FUENTE_TITULO)
        tit.setStyleSheet("border: none;")

        btnRetiro = Estilos.botonPeligro("↑ Retiro / Pago")
        btnRetiro.setFixedSize(160, 42)
        btnRetiro.clicked.connect(self.abrirDialogoRetiro)

        h_layout.addWidget(tit)
        h_layout.addStretch()
        h_layout.addWidget(btnRetiro)

        layout.addWidget(header)

        tarjetas = QWidget()
        tarjetas.setFixedHeight(110)
        t_layout = QHBoxLayout(tarjetas)
        t_layout.setSpacing(16)
        t_layout.setContentsMargins(0, 0, 0, 0)

        cardFondo, self.lblEfectivo = self.tarjetaCaja("Total Ventas", "$0.00", Estilos.BG_BLANCO, Estilos.TEXTO_PRINCIPAL)
        cardEfectivo, self.lblEfectivoTotal = self.tarjetaCaja("Efectivo en Caja", "$0.00", Estilos.INDIGO, QColor("white"))

        t_layout.addWidget(cardFondo)
        t_layout.addWidget(cardEfectivo)

        layout.addWidget(tarjetas)

        self.tablaMovs = QTableWidget(0, 4)
        self.tablaMovs.setHorizontalHeaderLabels(["Hora", "Tipo", "Concepto", "Monto"])
        self.tablaMovs.setFont(Estilos.FUENTE_NORMAL)
        self.tablaMovs.horizontalHeader().setFont(Estilos.FUENTE_XS)
        self.tablaMovs.horizontalHeader().setStyleSheet(f"background-color: {Estilos.BG_ZINC_100.name()};")
        self.tablaMovs.setStyleSheet(f"background-color: {Estilos.BG_BLANCO.name()}; border: 1px solid {Estilos.BORDE.name()};")
        self.tablaMovs.setEditTriggers(QTableWidget.EditTrigger.NoEditTriggers)
        self.tablaMovs.horizontalHeader().setSectionResizeMode(2, QHeaderView.ResizeMode.Stretch)
        self.tablaMovs.setColumnWidth(0, 100)
        self.tablaMovs.setColumnWidth(1, 110)
        self.tablaMovs.setColumnWidth(3, 120)

        layout.addWidget(self.tablaMovs)

        return p

    def tarjetaCaja(self, label, valor, bg, fg):
        card = QWidget()
        card.setStyleSheet(f"background-color: {bg.name()}; border: 1px solid {Estilos.BORDE.name() if bg == Estilos.BG_BLANCO else bg.name()}; border-radius: 4px;")
        layout = QVBoxLayout(card)
        layout.setContentsMargins(22, 18, 22, 18)
        
        lLbl = QLabel(label.upper())
        lLbl.setFont(Estilos.FUENTE_XS)
        lLbl.setStyleSheet(f"color: {'#C7D2FE' if bg == Estilos.INDIGO else Estilos.TEXTO_TENUE.name()}; border: none;")
        
        lVal = QLabel(valor)
        lVal.setFont(QFont("SansSerif", 34, QFont.Weight.Bold))
        lVal.setStyleSheet(f"color: {fg.name()}; border: none;")
        
        layout.addWidget(lLbl)
        layout.addWidget(lVal)
        
        return card, lVal

    def abrirDialogoRetiro(self):
        from Vista.RetiroDialog import RetiroDialog
        dlg = RetiroDialog(self.ventana.getFrame(), self.ctrl)
        dlg.exec()
        self.refrescar()

    def refrescar(self):
        if self.ctrl.isCajaAbierta():
            self.cardLayout.setCurrentIndex(self.TAB_ABIERTA)
            totalGeneral = sum(m.getMonto() for m in self.ctrl.getMovimientos() if m.getTipo() == TipoMovimiento.VENTA)
            
            if hasattr(self, 'lblEfectivo'):
                self.lblEfectivo.setText(f"${totalGeneral:.2f}")
            if hasattr(self, 'lblEfectivoTotal'):
                self.lblEfectivoTotal.setText(f"${self.ctrl.getEfectivoEsperado():.2f}")
                
            self.actualizarTablaMovimientos()
        else:
            self.cardLayout.setCurrentIndex(self.TAB_CERRADA)

    def actualizarTablaMovimientos(self):
        movs = list(reversed(self.ctrl.getMovimientos()))
        self.tablaMovs.setRowCount(len(movs))
        for i, m in enumerate(movs):
            fecha = m.getFecha().strftime("%H:%M:%S") if m.getFecha() else ""
            self.tablaMovs.setItem(i, 0, QTableWidgetItem(fecha))
            self.tablaMovs.setItem(i, 1, QTableWidgetItem(m.getTipo().name))
            self.tablaMovs.setItem(i, 2, QTableWidgetItem(m.getDescripcion()))
            monto_str = f"+${m.getMonto():.2f}" if m.esIngreso() else f"-${m.getMonto():.2f}"
            self.tablaMovs.setItem(i, 3, QTableWidgetItem(monto_str))

File: Vista/CajaPanel.cs
using Microsoft.Maui.Controls;
using Microsoft.Maui.Graphics;
using System;
using System.Collections.Generic;
using System.Linq;
using punto_de_venta_C_.Modelo;
using punto_de_venta_C_.Controlador;

namespace punto_de_venta_C_.Vista
{
    public class CajaPanel : ContentView
    {
        private readonly CajaControlador ctrl;
        private readonly Usuario usuario;
        private readonly VentanaPrincipal ventana;

        private Grid contenido;
        private View panelCerrada;
        private View panelAbierta;

        private Label lblEfectivo;
        private Label lblEfectivoTotal;
        private VerticalStackLayout movsLayout;

        public CajaPanel(CajaControlador ctrl, Usuario usuario, VentanaPrincipal ventana)
        {
            this.ctrl = ctrl;
            this.usuario = usuario;
            this.ventana = ventana;
            BackgroundColor = Estilos.BG_CLARO;
            Padding = new Thickness(24);
            construir();
        }

        private void construir()
        {
            contenido = new Grid();

            panelCerrada = construirPanelCerrada();
            panelAbierta = construirPanelAbierta();

            contenido.Add(panelCerrada);
            contenido.Add(panelAbierta);

            Content = contenido;
            refrescar();
        }

        private View construirPanelCerrada()
        {
            var grid = new Grid
            {
                BackgroundColor = Estilos.BG_CLARO,
                HorizontalOptions = LayoutOptions.Center,
                VerticalOptions = LayoutOptions.Center
            };

            var tarjeta = new VerticalStackLayout
            {
                BackgroundColor = Estilos.BG_BLANCO,
                WidthRequest = 420,
                HeightRequest = 440,
                Padding = new Thickness(40, 44),
                Spacing = 16
            };

            var icono = new Label
            {
                Text = "💵",
                FontSize = 64,
                HorizontalOptions = LayoutOptions.Center
            };
            tarjeta.Add(icono);

            var tit = new Label
            {
                Text = "Terminal Cerrada",
                FontSize = Estilos.FUENTE_TITULO_SIZE,
                FontAttributes = FontAttributes.Bold,
                HorizontalOptions = LayoutOptions.Center
            };
            tarjeta.Add(tit);

            var resp = new Label
            {
                Text = "Responsable: " + usuario.getNombre(),
                FontSize = Estilos.FUENTE_SMALL_SIZE,
                TextColor = Estilos.TEXTO_SECUNDARIO,
                HorizontalOptions = LayoutOptions.Center
            };
            tarjeta.Add(resp);

            tarjeta.Add(new BoxView { HeightRequest = 12, Color = Colors.Transparent });

            var lblFondoTexto = new Label
            {
                Text = "FONDO INICIAL ($)",
                FontSize = Estilos.FUENTE_XS_SIZE,
                TextColor = Estilos.TEXTO_TENUE
            };
            tarjeta.Add(lblFondoTexto);

            var txtFondo = new Entry
            {
                Text = "0.00",
                FontSize = 26,
                FontAttributes = FontAttributes.Bold,
                BackgroundColor = Estilos.BG_ZINC_100,
                TextColor = Estilos.TEXTO_PRINCIPAL,
                HeightRequest = 56
            };
            tarjeta.Add(txtFondo);

            tarjeta.Add(new BoxView { HeightRequest = 8, Color = Colors.Transparent });

            var btnAbrir = Estilos.botonPrimario("ABRIR TERMINAL");
            btnAbrir.FontSize = 16;
            btnAbrir.HeightRequest = 56;
            btnAbrir.Clicked += (s, e) =>
            {
                try
                {
                    double fondo = double.Parse(txtFondo.Text?.Trim() ?? "0");
                    ctrl.abrirCaja(fondo);
                }
                catch (Exception)
                {
                    ctrl.abrirCaja(0);
                }
            };
            tarjeta.Add(btnAbrir);

            grid.Children.Add(tarjeta);
            return grid;
        }

        private View construirPanelAbierta()
        {
            var layout = new Grid
            {
                RowDefinitions = new RowDefinitionCollection
                {
                    new RowDefinition { Height = GridLength.Auto },
                    new RowDefinition { Height = 16 },
                    new RowDefinition { Height = GridLength.Star }
                }
            };

            var header = new Grid
            {
                BackgroundColor = Estilos.BG_BLANCO,
                Padding = new Thickness(24, 20),
                ColumnDefinitions = new ColumnDefinitionCollection
                {
                    new ColumnDefinition { Width = GridLength.Star },
                    new ColumnDefinition { Width = 160 }
                }
            };

            var tit = new Label
            {
                Text = "GESTIÓN DE CAJA  —  Operativa",
                FontSize = Estilos.FUENTE_TITULO_SIZE,
                FontAttributes = FontAttributes.Bold,
                VerticalOptions = LayoutOptions.Center
            };
            header.Add(tit, 0, 0);

            var btnRetiro = Estilos.botonPeligro("↑ Retiro / Pago");
            btnRetiro.HeightRequest = 42;
            btnRetiro.Clicked += (s, e) => abrirDialogoRetiro();
            header.Add(btnRetiro, 1, 0);

            layout.Add(header, 0, 0);

            var centro = new Grid
            {
                RowDefinitions = new RowDefinitionCollection
                {
                    new RowDefinition { Height = 110 },
                    new RowDefinition { Height = 16 },
                    new RowDefinition { Height = GridLength.Star }
                }
            };

            var tarjetasGrid = new Grid
            {
                ColumnDefinitions = new ColumnDefinitionCollection
                {
                    new ColumnDefinition { Width = GridLength.Star },
                    new ColumnDefinition { Width = 16 },
                    new ColumnDefinition { Width = GridLength.Star }
                }
            };

            var cardFondo = tarjetaCaja("Total Ventas", "$0.00", Estilos.BG_BLANCO, Estilos.TEXTO_PRINCIPAL, out lblEfectivo);
            tarjetasGrid.Add(cardFondo, 0, 0);

            var cardEfectivo = tarjetaCaja("Efectivo en Caja", "$0.00", Estilos.INDIGO, Colors.White, out lblEfectivoTotal);
            tarjetasGrid.Add(cardEfectivo, 2, 0);

            centro.Add(tarjetasGrid, 0, 0);

            movsLayout = new VerticalStackLayout();
            var scroll = new ScrollView
            {
                Content = movsLayout,
                BackgroundColor = Estilos.BG_BLANCO
            };

            centro.Add(scroll, 0, 2);
            layout.Add(centro, 0, 2);

            return layout;
        }

        private Frame tarjetaCaja(string label, string valor, Color bg, Color fg, out Label labelRef)
        {
            var valLbl = new Label
            {
                Text = valor,
                FontSize = 34,
                FontAttributes = FontAttributes.Bold,
                TextColor = fg
            };
            labelRef = valLbl;

            var card = new Frame
            {
                BackgroundColor = bg,
                BorderColor = bg == Estilos.BG_BLANCO ? Estilos.BORDE : bg,
                CornerRadius = 8,
                Padding = new Thickness(22, 18),
                Content = new VerticalStackLayout
                {
                    Spacing = 6,
                    Children =
                    {
                        new Label
                        {
                            Text = label.ToUpper(),
                            FontSize = Estilos.FUENTE_XS_SIZE,
                            TextColor = bg == Estilos.INDIGO ? Color.FromRgb(199, 210, 254) : Estilos.TEXTO_TENUE
                        },
                        valLbl
                    }
                }
            };

            return card;
        }

        private async void abrirDialogoRetiro()
        {
            var dlg = new RetiroDialog(ctrl);
            await dlg.mostrar();
            refrescar();
        }

        public void refrescar()
        {
            if (ctrl.isCajaAbierta())
            {
                panelCerrada.IsVisible = false;
                panelAbierta.IsVisible = true;

                double totalGeneral = ctrl.getMovimientos()
                    .Where(m => m.getTipo() == Movimiento.Tipo.VENTA)
                    .Select(m => m.getMonto()).Sum();

                if (lblEfectivo != null)
                {
                    lblEfectivo.Text = string.Format("${0:0.00}", totalGeneral);
                }

                if (lblEfectivoTotal != null)
                {
                    lblEfectivoTotal.Text = string.Format("${0:0.00}", ctrl.getEfectivoEsperado());
                }

                actualizarTablaMovimientos();
            }
            else
            {
                panelCerrada.IsVisible = true;
                panelAbierta.IsVisible = false;
            }
        }

        private void actualizarTablaMovimientos()
        {
            if (movsLayout == null)
                return;

            var movs = new List<Movimiento>(ctrl.getMovimientos());
            movs.Reverse();
            movsLayout.Children.Clear();

            var header = new Grid
            {
                BackgroundColor = Estilos.BG_ZINC_100,
                HeightRequest = 30,
                ColumnDefinitions = new ColumnDefinitionCollection
                {
                    new ColumnDefinition { Width = 100 },
                    new ColumnDefinition { Width = 110 },
                    new ColumnDefinition { Width = GridLength.Star },
                    new ColumnDefinition { Width = 120 }
                }
            };
            header.Add(new Label { Text = "Hora", FontAttributes = FontAttributes.Bold, VerticalOptions = LayoutOptions.Center, Margin = new Thickness(10, 0) }, 0);
            header.Add(new Label { Text = "Tipo", FontAttributes = FontAttributes.Bold, VerticalOptions = LayoutOptions.Center }, 1);
            header.Add(new Label { Text = "Concepto", FontAttributes = FontAttributes.Bold, VerticalOptions = LayoutOptions.Center }, 2);
            header.Add(new Label { Text = "Monto", FontAttributes = FontAttributes.Bold, VerticalOptions = LayoutOptions.Center }, 3);
            movsLayout.Children.Add(header);

            foreach (Movimiento m in movs)
            {
                var rowGrid = new Grid
                {
                    HeightRequest = 40,
                    ColumnDefinitions = new ColumnDefinitionCollection
                    {
                        new ColumnDefinition { Width = 100 },
                        new ColumnDefinition { Width = 110 },
                        new ColumnDefinition { Width = GridLength.Star },
                        new ColumnDefinition { Width = 120 }
                    }
                };

                rowGrid.Add(new Label { Text = m.getFecha().ToString("HH:mm:ss"), VerticalOptions = LayoutOptions.Center, Margin = new Thickness(10, 0) }, 0);
                rowGrid.Add(new Label { Text = m.getTipo().ToString(), VerticalOptions = LayoutOptions.Center }, 1);
                rowGrid.Add(new Label { Text = m.getDescripcion(), VerticalOptions = LayoutOptions.Center }, 2);

                var montoLbl = new Label
                {
                    Text = string.Format(m.esIngreso() ? "+${0:0.00}" : "-${0:0.00}", m.getMonto()),
                    TextColor = m.esIngreso() ? Estilos.EMERALD : Estilos.ROSE,
                    FontAttributes = FontAttributes.Bold,
                    VerticalOptions = LayoutOptions.Center
                };
                rowGrid.Add(montoLbl, 3);

                movsLayout.Children.Add(rowGrid);
            }
        }
    }
}
